Supported image formats

gthvn1 opened the pull request:
This PR implements the supported image format mechanism proposed in this design document: https://xapi-project.github.io/new-docs/design/sm-supported-image-formats/index.html

- It adds supported image formats to the SM object if the SM plugin specifies it in its `DRIVER_INFO`.
- When the information is available, you can select which image format to use as the destination during a VM or VDI migration.

This feature is particularly useful because XCP-ng is adding support for the Qcow2 format in SMAPI to allow _VDIs_ larger than 2TB. So in the near future (we're currently releasing the beta version), some _SRs_ will support multiple formats such as VHD, RAW, and Qcow2. 

With this patch, it becomes possible to migrate a VM with VHD disks on one SR to another SR with Qcow2 disks.  If an SM plugin does not provide information about the supported image formats, the behavior remains unchanged.

For more details see the [specification](https://xapi-project.github.io/new-docs/design/sm-supported-image-formats/index.html).

gthvn1 committed:
Allow selection of image format during migration

This patch allows specifying the destination format for individual VDIs
mapped to a destination SR. It adds a new parameter to `VM.migrate_send`
and `VM.assert_can_migrate` API. It also adds a new parameter to XE CLI.
The format to specify the image format is `image-format:<source VDI
UUID>=<destination image format>`. If the given image format cannot be
validated, an error is returned.

It also adds a new parameter to `VDI.pool-migrate`. This new parameter
allows to provide a string that is the destination format. This string is
used to check whether the destination SR supports the expected format. If
the check fails or cannot be performed due to missing information on the
destination SR, an error is returned.

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

gthvn1 committed:
Datamodel: add supported_image_format field to SM object

When running `xe sm-list params=all` you will now have the info of
supported image formats if the SM plugin specified it in its DRIVER_INFO.
The field is called `supported-image-formats`. If the plugin doesn't
provide the info the field will be empty.

This patch modifies the datamodel and add a new field to store this
information into the SM object.

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

lindig commented on ocaml/idl/datamodel.ml:
```diff
@@ -5118,6 +5118,10 @@ module SM = struct
+        ; field ~lifecycle:[] ~qualifier:DynamicRO
+            ~default_value:(Some (VSet [])) ~ty:(Set String)
+            "supported_image_formats"
+            "Image formats supported by the SR (VHD, RAW, Qcow2, ...)"
```
Is the capitalisation significant? Would be good to list the supported formats in the exact syntax. The … leaves some imagination.

psafont replied:
> That said, I'm not sure we want to enfoce any format in XAPI, since someone could implement other formats such as vhdx, vmdk, or anything else.

Andriy spent quite a lot of time to be able to handle qcow2 in the toolstack. I don't think it's going to be possible to support all those image formats without intervention in xapi, and hence why it makes sense to code them in xapi. Then it can act as a gatekeeper, ensuring all allowed formats are properly supported

gthvn1 replied:
I agree that those image formats most probably won't  be supported without the help of XAPI. So I understand that it makes sense to list the supported format at the XAPI level. I will add the list of supported backends.

lindig commented on ocaml/idl/datamodel.ml:
```diff
@@ -5439,6 +5443,10 @@ module VDI = struct
+        ; ( String
+          , "dest_img_format"
+          , "The image format to use on destination SR"
+          )
```
Again, list supported formats in the exact syntax.

lindig commented on ocaml/tests/common/test_common.ml:
```diff
@@ -354,11 +354,12 @@ let make_sm ~__context ?(ref = Ref.make ()) ?(uuid = make_uuid ())
     ?(host_pending_features = []) ?(configuration = []) ?(other_config = [])
-    ?(driver_filename = "/dev/null") ?(required_cluster_stack = []) () =
+    ?(driver_filename = "/dev/null") ?(required_cluster_stack = [])
+    ?(supported_image_formats = []) () =
```
Is this a sensible default? Does this not force to set this parameter? Or is the empty list a special value?

lindig commented on ocaml/xapi-cli-server/cli_operations.ml:
```diff
@@ -4896,12 +4901,24 @@ let vm_migrate printer rpc session_id params =
+              let vdi =
+                Client.VDI.get_by_uuid ~rpc ~session_id ~uuid:vdi_uuid
+              in
+              debug "GTNDEBUG: add image format %s,%s" vdi_uuid vdi_fmt ;
```
What is `GTNDEBUG`?

gthvn1 commented:
> These appear to be some mysterious strings but the user of neither API or CLI should have to guess them,

Yes, in my mind the user should first check using `SM.get_all_records` to get the list, and then, when doing a migration, pick a format form that list.

psafont commented on ocaml/xapi-cli-server/record_util.ml:
```diff
@@ -566,3 +566,14 @@ let vm_placement_policy_of_string a =
   | s ->
       record_failure "Invalid VM placement policy, got %s" s
+
+let image_format_type_of_string s =
```
I don't think this function belongs in this module, see the comment at the top:

```
(* NOTE: Unless conversion requires some custom logic, no new functions should
   be added here. Automatically-generated functions with consistent behaviour
   and naming are generated from the datamodel and included here.
```

psafont commented on ocaml/xapi-cli-server/records.ml:
```diff
@@ -3867,7 +3875,10 @@ let sm_record rpc session_id sm =
       ; make_field ~name:"supported-image-formats"
           ~get:(fun () ->
-            concat_with_comma (x ()).API.sM_supported_image_formats
+            concat_with_comma
```
Please don't concatenate with commas, see the comment for the function:

```
(* Some lists are separated with commas and cannot be changed to preserve
   backwards compatibility. Do not use this function on new fields. *)
```

Use concat_with_semi instead

In this case, you might want to use `map_and_concat`

gthvn1 commented on ocaml/idl/datamodel.ml:
```diff
@@ -4998,6 +4998,16 @@ module SR = struct
 end
 
 module SM = struct
+  let image_format_type =
```
And I'm not sure to see how "formats" is then use as "~ty:" of the field. I should put "(Map (String,String))" right?

gthvn1 replied:
Also I agree to use `formats'` in line 5134 but what about line 5457 for destination image. Expected formats for module VDI are the same so should I duplicate the definition of `formats'` for VDI?

gthvn1 replied:
In fact what I don't understand is that "image_format_type" is a type that can be used for `~ty` but 'formats'is not. So I'm little bit lost here because for me it is not the same meaning. Or should I create an Enum from "formats"?

gthvn1 committed:
Add new parameter to VM.migrate_send in GO SDK

Update VM.MigrateSend call to include new VdiFormatMap parameter.

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

gthvn1 commented:
To keep you informed I almost successfully run our CI with this patch installed... Almost because in fact there is an issue with smapiv3 drivers. I focused on v1 but it looks like v3 is expecting a specific query_result (https://xapi-project.github.io/xapi-storage/#plugin-type-definitions) and in my case I added a new field on xapi side and I probably missed a default value for v3. I have one for v1. But I'm not sure if we can have optional field in the query_result.
I'm currently reading & trying to understand smapiv3 stuff.

gthvn1 commented on ocaml/xapi-storage-script/main.ml:
```diff
@@ -936,6 +936,8 @@ module QueryImpl (M : META) = struct
         ; required_cluster_stack=
             response.Xapi_storage.Plugin.required_cluster_stack
+        ; supported_image_formats=
+            response.Xapi_storage.Plugin.supported_image_formats
```
In fact if I understand correctly the query_result is the same for all smapi version. So I can't really remove it from smapiv3. I will check how to ignore it properly...

gthvn1 committed:
Bumping database schema version

A new field supported_image_format and new parameters have been added for:
  - VM.migrate_send
  - VM.assert_can_migrate
  - VDI.pool_migrate

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

psafont commented on ocaml/idl/datamodel_vm.ml:
```diff
@@ -1714,6 +1714,13 @@ let migrate_send =
+          param_type= Map (Ref _vdi, String)
+        ; param_name= "vdi_format_map"
+        ; param_doc= "Map of source VDI to it's expected type on destination"
+        ; param_release= numbered_release "25.39.0-next"
```
```suggestion
        ; param_release= numbered_release "26.0.0-next"
```

psafont commented on ocaml/xapi-storage-script/main.ml:
```diff
@@ -1817,7 +1817,9 @@ module DATAImpl (M : META) = struct
-  let mirror dbg sr vdi' vm' remote =
+  let mirror dbg sr vdi' image_format vm' remote =
+    let _ = image_format in
+    (* TODO: really use image format *)
```
Is this still missing?

psafont reviewed (commented):
I thought this would be almost ready to go, but I've seen that there's a TODO, and the generator for smapiv3 plugins has not been modified, which is not what I was expecting, can you expand on those? The latter may be left for later because it shouldn't break anything, just not add nice support for the feature in smapiv3. but I don't understand the implication in the mirror call

psafont commented on ocaml/idl/datamodel_common.ml:
```diff
@@ -10,7 +10,7 @@ open Datamodel_roles
 let schema_major_vsn = 5
 
-let schema_minor_vsn = 792
+let schema_minor_vsn = 793
```
Not after this is merged: https://github.com/xapi-project/xen-api/pull/6820/changes ;)

gthvn1 committed:
quicktest: add VDI migration test

Introduce a new quicktest covering local VDI migration between two
Storage Repositories (SRs).

Add a `migration_path` filter that injects a `(src, dst)` SR pair
derived from an `SR.srs` constraint. The filter selects a single
valid migration path (if available) and generates one test case.
If fewer than two compatible SRs exist, no test is produced.

The test:
- Creates a VDI on the source SR
- Attaches it to a temporary VM
- Calls VDI.pool_migrate to the destination SR
- Verifies that the VDI's SR has changed accordingly
- Cleans up safely, tracking ownership transfer when migration
  replaces the original VDI

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

gthvn1 commented:
- Rebased on master
- Removed `supported_image_format` from the query result in the plugin generator because, currently, even a default value forces the plugin to implement it. The generated code makes the field mandatory and that breaks current SMAPIv3 plugin. We have proposed a patch to fix this: https://github.com/mirage/ocaml-rpc/pull/190 

As there are no SMAPIv3 plugins that support live migration, we are using an empty list for now, and it cannot be modified at this time.

gthvn1 commented:
- Fix schema hash

psafont commented on ocaml/xapi/storage_smapiv3_migrate.ml:
```diff
@@ -189,18 +190,22 @@ module MIRROR : SMAPIv2_MIRROR = struct
           )
     )
 
-  let receive_start _ctx ~dbg:_ ~sr:_ ~vdi_info:_ ~id:_ ~similar:_ =
```
Are you sure it's safe to modify calls like this? Could you check the reason there are 3 different receive_start calls? We might need newer ones. I think it might be related to backwards compatibility when doing RPUs

gthvn1 replied:
Yes I will check. But you are correct that there are different flavor for backwards compatibility.

gthvn1 replied:
After testing migrations from a host with a supported image format to another host (in a different pool) that does not support it (in both directions) I didn’t encounter any issues.

 As I understand it, `receive_start` is called on the source and prepares the mirroring. In the end, it sends a `VDI.create` XML-RPC call (and some others) to the destination, but I don’t see any issue with `receive_start` itself.

The process will fail if you explicitly try to use the image format, because the source checks whether the destination SR uses an SM type that supports that format. However, if the option is not set, an empty list is used and no checks are performed against the destination SR, so it should work.

That said, even if I think that it is safe to modify the call, I agree that it doesn’t really make sense for a deprecated API.

I’ll also check intra-pool migration, but as far as I remember, VDI migration behaves the same way.

gthvn1 replied:
So it looks like there is an issue if you want to do a RPU (but it is not related to `receive_start`). The scenario is with two host A (master) and B:
1. vdi pool migrate from A to B (ok)
2. update A (so now A supports image format)
3. reboot A
4. we want to update B so we need to move the VM back to A but:
```
# xe vdi-pool-migrate uuid=4701fa33-c63d-49ff-baad-80e008dd6ab3 sr-uuid=21c7d26b-1ce7-04ea-1685-ba45e50c4191
You tried to call a method with the incorrect number of parameters. The fully-qualified method name that you used, and the number of received and expected parameters are returned.
method: VDI.pool_migrate
expected: 3
received: 4
```
There is an issue with the database because the new field is only available on the new host so B complains.
```
Apr 22 11:01:45 xcp-gtn-ip13 xapi: [ warn||326 HTTPS 10.1.38.12->:::80|host.request_backup D:d98b24f2a7cf|Xapi_database__Db_xml] no lifetime information about SM.supported_image_formats, ignoring
Apr 22 11:01:45 xcp-gtn-ip13 xapi: [ warn||326 HTTPS 10.1.38.12->:::80|host.request_backup D:d98b24f2a7cf|xmlrpc_client] stunnel pid: 347171 caught Db_exn.DBCache_NotFound("missing column", "SM", "supported_image_formats")
Apr 22 11:01:45 xcp-gtn-ip13 xapi: [error||326 :::80|dispatch:host.request_backup D:6671e841a40d|backtrace] host.request_backup D:d98b24f2a7cf failed with exception Db_exn.DBCache_NotFound("missing column", "SM", "supported_image_formats")
Apr 22 11:01:45 xcp-gtn-ip13 xapi: [error||326 :::80|dispatch:host.request_backup D:6671e841a40d|backtrace] Raised Db_exn.DBCache_NotFound("missing column", "SM", "supported_image_formats")
Apr 22 11:01:45 xcp-gtn-ip13 xapi: [error||326 :::80|dispatch:host.request_backup D:6671e841a40d|backtrace] 1/21 xapi Raised at file ocaml/database/schema.ml, line 190
```
I'm investigating that...

gthvn1 replied:
See the [comment below](https://github.com/xapi-project/xen-api/pull/6712#pullrequestreview-4154480333) for the explanation

gthvn1 commented on ocaml/xapi/message_forwarding.ml:
```diff
@@ -5483,7 +5484,7 @@ functor
         )
 
-      let pool_migrate ~__context ~vdi ~sr ~options =
+      let pool_migrate ~__context ~vdi ~sr ~dest_img_format ~options =
```
This is the issue with RPU scenario. The trick is that the VDI.pool_migrate is not executed on master but forwarded to the expected host. So once master has been updated (and slaves are not) the XML-RPC is built using the new format that has four parameters. But the slave expects only 3 parameters.
So the problem is that this particular call is forwarded. 
So I guess that we should put the "dest_img_format" into ~options otherwise RPU is broken

gthvn1 commented:
For information I'm currently running these scenarios:
  - **Cross pool migration**
    - Without using the option supported image format (and without QCow2 support)
      - [x] migration from A (with patches) to B (no patches): `xe vm-migrate ...`
      - [x] migration from B (no patches) to A (with patches): `xe vm-migrate ...`
      - [x] migration from A (with patches) to B (with patches): `xe vm-migrate ...`
    - Using the option supported image format (+ QCow2 support)
      - [x] migration from B VHD to A (currently it generates a QCow2 file): `xe vm-migrate ...`
      - [x] migration from A QCow2 to B QCow2: `xe vm-migrate ...`
      - [x] migration from B QCow2 to A VHD: `xe vm-migrate ...image-format:<VDIUUID>=vhd`
      - [ ] ~~migration from A VHD to B RAW~~: `xe vm-migrate ...image-format:<VDIUUID>=raw`
        - Failed (see below the error) 
      - [ ] ~~migration from A Qcow2 to B RAW~~: `xe vm-migrate ...image-format:<VDIUUID>=raw`
        - Failed (same as above) 
      - [ ] ~~migration from B RAW to A Qcow2~~: `xe vm-migrate ...image-format:<VDIUUID>=qcow2`
  - **Intra pool migration (simulate RPU)**
    - Without using the option supported image format (no qcow2 support)
      - [x] migration from A1 (no patches) to A2 (no patches): "evacuate" master (xe vdi-pool-migrate because I don't have shared SR so not real evacuate)
      - [x] migration from A2 (no patches) to A1 (with patches): `xe vdi-pool-migrate ...`
      - [x] migration from A1 (with patches) to A2 (with patches): `xe vdi-pool-migrate ...`
    - Using the option supported image format (+ QCow2 support)
      - [x] migration from A2 VHD to A1 (currently it generates a QCow2 file): `xe vdi-pool-migrate ...`
      - [x] migration from A1 QCow2 to A2 QCow2: `xe vdi-pool-migrate ...`
      - [x] migration from A2 QCow2 to A1 VHD: `xe vdi-pool-migrate ... image-format:<VDI_UUID>=vhd`
      - [x] migration from A1 VHD to A2 Qcow2: `xe vdi-pool-migrate ... image-format:<VDI_UUID>=qcow2`
      - [ ] ~~migration from A2 QCow2 to A1 RAW~~: `xe vdi-pool-migrate ... image-format:<VDI_UUID>=raw`
        - Failed (same as above)
  - **quicktest** (newly added in this PR):
    - [x] smapiv1 migration
    - [ ] ~~smapiv3 migration~~: cannot be tested because we don't have any plugin that supports migration
  - **Expected failures** (check is done before migration):
    - [x] passing a wrong parameter to `vdi-pool-migrate`: `message: Storage_error ([S(Internal_error);S(Storage_error ([S(Migration_preparation_failure);S(Storage_error ([S(Backend_error);[S(SR_BACKEND_FAILURE_66);[S();S(Invalid VDI type [opterr=Invalid VDI type None]);S()]]]))]))])`
    - [x] passing a wrong parameter to `vm-migrate`
    - [x] passing a parameter that is not in the list of supported_image_format
      - [x] vdi pool migrate: Got `vdi: Image format qcow2 is not supported by XXX` where list is ["raw"; "vhd"]
      - [x] vm migrate: Got `vdi: Image format qcow2 is not supported by XXX` where list is ["raw"; "vhd"]

- **NOTES**:
  - By default supported image format of SM is an empty list. In this case there is no check before migration
  - By QCow2 support I mean installing [xcp-ng qcow2](https://xcp-ng.org/forum/topic/10308/dedicated-thread-removing-the-2tib-limit-with-qcow2-volumes/19) 
  - `with patches` mean Host with this current PR applied
  - Tests are in this order because it allows to run them consecutively

- **FAILURES**:
  - extra and intra pool migration from qcow2 to raw.
    - *Note: this is not related to this PR since it also happens with qcow2 from `xcp-ng-testing`.*
```sh
# xe vdi-pool-migrate uuid=882cedbd-e8ef-4c5d-9a57-b99b91516b49  sr-uuid=d835d019-1b97-0080-1bc0-eb52542db3f5 dest-img-format=raw
The server failed to handle your request, due to an internal error. The given message may give details useful for debugging the problem.
message: Storage_error ([S(Internal_error);S(Storage_error ([S(Migration_preparation_failure);S(Storage_error ([S(Unimplemented);S(VDI.snapshot)]))]))])
```

gthvn1 commented on ocaml/xapi-cli-server/cli_frontend.ml:
```diff
@@ -2480,10 +2483,10 @@ let rec cmdtable_data : (string * cmd_spec) list =
-          "Migrate a VDI to a specified SR, while the VDI is attached to a \
-           running guest."
+          "Migrate a VDI to a specified SR, while it is attached to a running \
+           guest. You can specify the image format for the destination."
```
Add what format is allowed: RAW, VHD and QCow2

gthvn1 commented on ocaml/idl/datamodel.ml:
```diff
@@ -5445,7 +5460,11 @@ module VDI = struct
-        ; (Map (String, String), "options", "Other parameters")
+        ; ( Map (String, String)
+          , "options"
+          , "Extra parameters. Supports: \"dest-img-format\" (raw|vhd|qcow2) \
```
We probably want to be consistent with VM.pool-migrate that uses "img-format" ?

gthvn1 commented:
- I have passed all tests described [here](https://github.com/xapi-project/xen-api/pull/6712#issuecomment-4299109107)
  - everything looks good 
- The failures we are seeing concern RAW migration, but it appears to be a consistent issue with the current version of XCP-ng.

psafont reviewed (approved):
With the battery of tests run by Guillaume, I think this is ready to be merged
